### PR TITLE
Add world storage for player ranks

### LINKED ISSUE #2: Doesn't work for offline players

### PR DESCRIPTION
**Pull request by An0n3m0us**
Closes: #2 

Stores a file in each world folder (`ranks.txt`) which contains players names and ranks. It works where once the player has joined and a rank has been assigned, the rank can always be changed whether the player is online or offline. But if the rank of the player is cleared, they will have to be 're-ranked' while online.

Of course, currently existing ranks for players will not exist in the file as I'm not sure how I'd get them without all of the players being online; meaning that from now on, the ranks will be stored in a file.

### CODE REVIEW

**Comment by octacian** [comment]
Hey, first off, thanks for the PR! However, rather than review it, maybe I'll just share my thoughts here since they largely apply to the entire change.

I'm glad this was brought to my attention again, as this is definitely a feature that would be useful. However, I'm not sure that a raw text file is the best way to go about implementing it. My first thought was to suggest utilizing the Minetest mod storage API, but then something else occurred to me. xban2 and similar mods also have to keep track of offline players and almost every one of them has long since taken the route of storing information in an SQL database due to performance issues otherwise incurred. Now, in the case of ranks being used only to differentiate server staff or other low-population classes, such a thing should not be a concern. However, what if every player was given a rank? Could the performance of text-based save files, including the mod storage API, then be a concern if used on larger servers?

I might be overthinking this though, so I would love to hear your thoughts on the issue.

**Comment by An0n3m0us** [comment]
I completely agree with using a database, it's probably much more efficient and cleaner than storing stuff in a text file; I'm new to mod storage, so I lack knowledge in knowing what causes performance issues.

As you said, the best way would be storing/loading it from a database in each world.
I'll have a look into the xban2 mod code and see if I can improve my PR.

**Comment by An0n3m0us** [comment]
I've added a DB process using the xban2 code.

**Comment by octacian** [comment]
Alright so I did some research and talked to some friends and it looks like I was wrong, not only does xban2 not actually use an SQL database (it's sban that uses SQLite) but it probably isn't the best option for ranks. The reason for this is that for SQLite to be utilized by a mod, it has to be installed through LuaRocks, and if LuaRocks isn't present, it has to be installed as well. And then the mod has to be added as trusted.

So, tl;dr, the best option, at least until someone *actually* runs into performance issues (assuming it ever happens), appears to be mod storage.

Sorry for misleading! Thank you for your effort so far, if you want to refactor this to use mod storage I would be happy to review and merge it afterward.

**Comment by An0n3m0us** [comment]
Yeah. I did some research and I found out that text files can actually be quite fast if used in the correct way.
I'll work on it soon!

**Comment by An0n3m0us** [comment]
There; that should do it. I've added mod storage and tested it too.

File: init.lua [hunk 1]
```diff
@@ -52,12 +56,8 @@ end
 
 -- [function] Get player rank
 function ranks.get_rank(player)
-	if type(player) == "string" then
-		player = minetest.get_player_by_name(player)
-	end
```
Comment (octacian) [inline init.lua]: Some variation of this snippet should be kept. With your changes, the value stored in the `player` variation should always be a string. Previously this snippet ensured that the value was always an ObjectRef, fetching one if the value was a string. In this case, it should be reversed to fetch the player name from the ObjectRef via `player:get_player_name()` if the `player` variable was not a string. It might also make sense to change all instances of `player` to `name` in context with the above considerations.

File: init.lua [hunk 2]
```diff
@@ -82,8 +82,8 @@ function ranks.update_privs(player, trigger)
 	end
 
 	local name = player:get_player_name()
-	local rank = ranks.get_rank(player)
-	if rank then
+	local rank = storage:get_string(name)
+	if rank ~= "" then
```
Comment (octacian) [inline init.lua]: `ranks.get_rank` works as a wrapper for any storage mechanism, and as such, there is no need to change anywhere that the function is called. Since it still returns under the same conditions, it may still be used under the same conditions. In other words, no change at all should be necessary here.

File: init.lua [hunk 3]
```diff
@@ -157,8 +157,8 @@ function ranks.update_nametag(player)
...
```
Comment (octacian) [inline init.lua]: Again, no change should be necessary here.

File: init.lua [hunk 4]
```diff
@@ -177,53 +177,64 @@ function ranks.update_nametag(player)
...
+		storage:set_string(name, rank)
+
 		-- Set attribute
-		player:set_attribute("ranks:rank", rank)
-		-- Update nametag
-		ranks.update_nametag(player)
-		-- Update privileges
-		ranks.update_privs(player)
+		if minetest.get_player_by_name(name) then
+			player:set_attribute("ranks:rank", rank)
```
Comment (octacian) [inline init.lua]: With attribute-based storage no longer in use, this is not necessary.

File: init.lua [hunk 5]
```diff
@@ -177,53 +177,64 @@ function ranks.update_nametag(player)
...
 end
 
--- [function] Set player rank
+-- [function] Set player rank TODO
 function ranks.set_rank(player, rank)
+	local name = player
+
 	if type(player) == "string" then
 		player = minetest.get_player_by_name(player)
 	end
```
Comment (octacian) [inline init.lua]: Similarly to lines -55 to -57, some variation of this check should be retained, except modified to ensure that the content of `player` is always a `string` rather than an ObjectRef.

File: init.lua [hunk 6]
```diff
@@ -177,53 +177,64 @@ function ranks.update_nametag(player)
...
 end
 
--- [function] Set player rank
+-- [function] Set player rank TODO
 function ranks.set_rank(player, rank)
+	local name = player
```
Comment (octacian) [inline init.lua]: Instead of creating a new variable, you may as well just rename `player` to `name` :man_shrugging:

File: init.lua [hunk 7]
```diff
@@ -177,53 +177,64 @@ function ranks.update_nametag(player)
...
+			ranks.update_privs(player)
+		end
 
 		return true
 	end
 end
 
 -- [function] Remove rank from player
 function ranks.remove_rank(player)
+	local name = player
```
Comment (octacian) [inline init.lua]: Again, may as well rename `player` rather than creating a new variable.

File: init.lua [hunk 8]
```diff
@@ -177,53 +177,64 @@ function ranks.update_nametag(player)
...
+
 	if type(player) == "string" then
 		player = minetest.get_player_by_name(player)
 	end
```
Comment (octacian) [inline init.lua]: As mentioned in the cases of `ranks.set_rank` and `ranks.get_rank`, this should be modified to ensure that the content of `player` is always a string.

File: init.lua [hunk 9]
```diff
@@ -177,53 +177,64 @@ function ranks.update_nametag(player)
...
-		player:set_nametag_attributes({
-			text = name,
-			color = "#ffffff",
-		})
-		-- Update privileges
-		local basic_privs =
-			minetest.string_to_privs(minetest.settings:get("basic_privs") or "interact,shout")
-		minetest.set_player_privs(name, basic_privs)
+	local rank = storage:get_string(name)
+	if rank ~= "" then
```
Comment (octacian) [inline init.lua]: `ranks.get_rank` may be utilized here and we may check if it is just `nil` via `if rank then` rather than having to check if `rank` is an empty string since `ranks.get` always returns either the name of the rank or `nil`.

File: init.lua [hunk 10]
```diff
@@ -177,53 +177,64 @@ function ranks.update_nametag(player)
...
+		storage:set_string(name, nil)
+
+		if minetest.get_player_by_name(name) then
+			-- Clear attribute
+			player:set_attribute("ranks:rank", nil)
```
Comment (octacian) [inline init.lua]: With attribute-based storage no longer in use, this is not necessary.

File: init.lua [hunk 11]
```diff
@@ -177,53 +177,64 @@ function ranks.update_nametag(player)
...
 	end
 end
 
 -- [function] Send prefixed message (if enabled)
 function ranks.chat_send(name, message)
 	if minetest.settings:get("ranks.prefix_chat") ~= "false" then
-		local rank = ranks.get_rank(name)
-		if rank then
+		local rank = storage:get_string(name)
+		if rank ~= "" then
```
Comment (octacian) [inline init.lua]: No change should be necessary here.

File: init.lua [hunk 12]
```diff
@@ -252,7 +263,7 @@ minetest.register_privilege("rank", {
 
 -- Assign/update rank on join player
 minetest.register_on_joinplayer(function(player)
-	if ranks.get_rank(player) then
+	if storage:get_string(player:get_player_name()) then
```
Comment (octacian) [inline init.lua]: No change should be necessary here due to retaining the check in `ranks.get_rank` that ensures that the argument passed in is always a string or if not, is converted to one.

File: init.lua [hunk 13]
```diff
@@ -315,15 +326,19 @@ minetest.register_chatcommand("getrank", {
 	params = "<name> | name of player",
 	func = function(name, param)
 		if param and param ~= "" then
-			if minetest.get_player_by_name(param) then
-				local rank = ranks.get_rank(param) or "No rank"
+			if storage:get_string(param) then
+				local rank = storage:get_string(param):gsub("^%l", string.upper)
```
Comment (octacian) [inline init.lua]: `ranks.get_rank` should again be suitable. Also, what does the `:gsub` call do? Does it make the first character uppercase or something? Sorry, I'm not familiar with Lua's take on regex :sweat_smile:

File: init.lua [hunk 14]
```diff
@@ -315,15 +326,19 @@ minetest.register_chatcommand("getrank", {
...
 			end
 		else
-			local rank = ranks.get_rank(name) or "No rank"
-			return false, "Your rank: "..rank
+			if storage:get_string(name) then
+				local rank = storage:get_string(name):gsub("^%l", string.upper) or "No rank"
+				return false, "Your rank: "..rank
+			else
+				return false, "Your rank: No rank"
+			end
```
Comment (octacian) [inline init.lua]: Does any of this really need to change thanks to `ranks.get_rank`?

File: init.lua [hunk 15]
```diff
@@ -6,6 +6,10 @@ local chat3_exists = minetest.get_modpath("chat3")
 local registered   = {}
 local default
 
+-- Load mod storage
+
```
Comment (octacian) [inline init.lua]: Alright so I'm being extremely, probably unnecessarily and maybe even a tiny bit obnoxiously picky right now, but this whitespace has really been annoying me :joy:

**Review by octacian — CHANGES_REQUESTED** [review]
Nice work on implementing and testing mod storage! I've looked over all the code and there's a few things that I think could be made more efficient, so I've left wayyy too many comments to express this :joy:

Oh and regarding existing ranks, they can actually be imported quite easily. There is a `minetest.register_on_joinplayer` callback near the bottom into which this functionality could be easily integrated. Simply check the player attributes for a rank, and if one is found, call `ranks.set_rank` to update the new mod storage "database" and finish off by removing the attribute. I'd recommend putting this all right at the start of the callback so that all later code can still be executed in the correct context.

File: init.lua (hunk 6 above)
Comment (An0n3m0us) [reply init.lua]: It's overwritten on line 185 I believe.

File: init.lua (hunk 15 above)
Comment (An0n3m0us) [reply init.lua]: xD I'll remove it!

File: init.lua (hunk 6 above)
Comment (octacian) [reply init.lua]: Yeah, you should be able to safely change all instances in which `player` is used to `name` though, including on line 185.

File: init.lua (hunk 2 above)
Comment (An0n3m0us) [reply init.lua]: I set it to `ranks.get_rank(name)`

File: init.lua (hunk 6 above)
Comment (An0n3m0us) [reply init.lua]: Will do!

File: init.lua (hunk 3 above)
Comment (An0n3m0us) [reply init.lua]: OK, done!

File: init.lua (hunk 4 above)
Comment (An0n3m0us) [reply init.lua]: Done!

File: init.lua (hunk 5 above)
Comment (An0n3m0us) [reply init.lua]: OK, will do!

File: init.lua (hunk 10 above)
Comment (An0n3m0us) [reply init.lua]: Done!

File: init.lua (hunk 12 above)
Comment (An0n3m0us) [reply init.lua]: Done!

File: init.lua (hunk 13 above)
Comment (An0n3m0us) [reply init.lua]: You are correct; it does just that : )
Neither am I xD I just searched around for a while until I found out how to do it.

Comment (An0n3m0us) [reply init.lua]: Done!

File: init.lua (hunk 6 above)
Comment (An0n3m0us) [reply init.lua]: Done!

File: init.lua (hunk 7 above)
Comment (An0n3m0us) [reply init.lua]: Done!

File: init.lua (hunk 9 above)
Comment (An0n3m0us) [reply init.lua]: Done!

File: init.lua (hunk 11 above)
Comment (An0n3m0us) [reply init.lua]: Done!

File: init.lua (hunk 15 above)
Comment (An0n3m0us) [reply init.lua]: Done!

File: init.lua (hunk 1 above)
Comment (An0n3m0us) [reply init.lua]: So like this?:

```
function ranks.get_rank(name)
	if type(name) == "string" then
		local rank = storage:get_string(name)
	else
		local rank = storage:get_string(minetest.get_player_by_name(name))
	end

	if rank ~= "" and registered[rank] then
		return rank
	end
end
```

Comment (octacian) [reply init.lua]: Hmmm I would probably go for something like this:
```lua
function ranks.get_rank(name)
	if type(name) ~= "string" then
		name = name:get_player_name()
	end
	
	local rank = storage:get_string(name)
	if rank ~= "" and registered[rank] then
		return rank
	end
end
```
You see, we are simply trying to ensure that the variable `name` is always a string. If it isn't a string, it's safe to assume that it is an ObjectRef (Lua will recognize it as the `userdata` type). And from the Lua API documentation, we know that we can fetch the raw string name of a player from an ObjectRef using `get_player_name()`. With all this in mind, we can simply reassign name with the string returned from `get_player_name`.

File: init.lua [hunk 16]
```diff
@@ -72,18 +74,17 @@ function ranks.get_def(rank)
...
 
 	if not player then
 		return
 	end
 
-	local name = player:get_player_name()
-	local rank = ranks.get_rank(player)
-	if rank then
+	local rank = ranks.get_rank(name)
+	if ranks.get_rank(name) ~= nil then
```
Comment (octacian) [inline init.lua]: Line 86 uses the variable `rank` to store the player's rank. There is no need to fetch the rank again on line 87, we need only reference the variable `rank`. The same applies in `update_nametag`, `set_rank` and `remove_rank`.

Comment (An0n3m0us) [reply init.lua]: Done!

File: init.lua (hunk 1 above)
Comment (An0n3m0us) [reply init.lua]: Okay!
I think it's done now.

File: init.lua (hunk 14 above)
Comment (An0n3m0us) [reply init.lua]: I'm working on this now! I've found a better way to do it, with less code.

Comment (An0n3m0us) [reply init.lua]: I've changed it.

File: init.lua [hunk 17]
```diff
@@ -72,18 +75,17 @@ function ranks.get_def(rank)
 end
 
 -- [function] Update player privileges
-function ranks.update_privs(player, trigger)
-	if type(player) == "string" then
-		player = minetest.get_player_by_name(player)
+function ranks.update_privs(name, trigger)
+	if type(name) == "string" then
+		player = minetest.get_player_by_name(name)
```
Comment (octacian) [inline init.lua]: This should be
```lua
if type(name) ~= "string" then
    name = name:get_player_name()
```
since only the string name is required below, not the entire player userdata object.

File: init.lua [hunk 18]
```diff
@@ -147,18 +148,17 @@ function ranks.update_privs(player, trigger)
...
-function ranks.update_nametag(player)
+function ranks.update_nametag(name)
 	if minetest.settings:get("ranks.prefix_nametag") == "false" then
 		return
 	end
 
-	if type(player) == "string" then
-		player = minetest.get_player_by_name(player)
+	if type(name) == "string" then
+		player = minetest.get_player_by_name(name)
```
Comment (octacian) [inline init.lua]: Again this should be effectively flipped:
```lua
if type(name) ~= "string" then
    name = name:get_player_name()
```

File: init.lua [hunk 19]
```diff
@@ -169,61 +169,63 @@ function ranks.update_nametag(player)
 			prefix = ""
 		end
 
-		player:set_nametag_attributes({
-			text = prefix..name,
-		})
+		if player then
```
Comment (octacian) [inline init.lua:171]: Unless I've missed something `player` isn't actually defined anywhere anymore. Define it above the opening clause of the if statement, e.g:
```lua
local player = minetest.get_player_by_name(name)
```

File: init.lua [hunk 20]
```diff
@@ -169,61 +169,63 @@ function ranks.update_nametag(player)
...
 end
 
--- [function] Set player rank
-function ranks.set_rank(player, rank)
-	if type(player) == "string" then
-		player = minetest.get_player_by_name(player)
+-- [function] Set player rank TODO
+function ranks.set_rank(name, rank)
+	if type(name) == "string" then
+		player = minetest.get_player_by_name(name)
```
Comment (octacian) [inline init.lua]: Again, this statement should be effectively flipped. See other comments to the same end.
Also, I see `TODO` in the comment above. Is something else still to be done or should that be removed?

File: init.lua [hunk 21]
```diff
@@ -169,61 +169,63 @@ function ranks.update_nametag(player)
...
 	end
 end
 
 -- [function] Remove rank from player
-function ranks.remove_rank(player)
-	if type(player) == "string" then
-		player = minetest.get_player_by_name(player)
+function ranks.remove_rank(name)
+	if type(name) == "string" then
+		name = minetest.get_player_by_name(name)
```
Comment (octacian) [inline init.lua]: Once again, this should be effectively flipped. See other comments to the same end.

File: init.lua [hunk 22]
```diff
@@ -169,61 +169,63 @@ function ranks.update_nametag(player)
...
-		local basic_privs =
-			minetest.string_to_privs(minetest.settings:get("basic_privs") or "interact,shout")
-		minetest.set_player_privs(name, basic_privs)
+	local rank = ranks.get_rank(name)
+	if rank ~= nil then
+		storage:set_string(name, nil)
+
+		if minetest.get_player_by_name(name) then
+			-- Update nametag
+			player:set_nametag_attributes({
```
Comment (octacian) [inline init.lua]: `player` isn't defined. Define it outside the first clause of the if statement and replace the conditional to be `if player then ... end`.

File: init.lua [hunk 23]
```diff
@@ -283,7 +287,7 @@ minetest.register_chatcommand("rank", {
 		if #param == 1 and param[1] == "list" then
 			return true, "Available Ranks: "..ranks.list_plaintext()
 		elseif #param == 2 then
-			if minetest.get_player_by_name(param[1]) then
+			if minetest.get_player_by_name(param[1]) or ranks.get_rank(param[1]) then
```
Comment (octacian) [inline init.lua]: We don't really need to check if `minetest.get_player_by_name` returns anything, `ranks.get_rank` does all the work for us at once.

**Review by octacian — CHANGES_REQUESTED** [review]
I'm sorry for taking so long to get back to you. Hopefully this will be the last set of changes necessary, I'll try to review them much more quickly this time :)

File: init.lua (hunk 18 above)
Comment (An0n3m0us) [reply init.lua]: See line *169*. Does this still need to be implemented in some form?

File: init.lua (hunk 17 above)
Comment (An0n3m0us) [reply init.lua]: Done!

File: init.lua (hunk 19 above)
Comment (An0n3m0us) [reply init.lua:171]: This is related to the review above.

File: init.lua (hunk 20 above)
Comment (An0n3m0us) [reply init.lua]: Done!
I've removed the TODO too; although I cannot remember why I put it there xD
It's been a long time since I've looked at the code.

But I tested it anyway; just to make sure it is working correctly.

File: init.lua (hunk 21 above)
Comment (An0n3m0us) [reply init.lua]: But `minetest.get_player_by_name(name)` != `name:get_player_name()`

File: init.lua (hunk 22 above)
Comment (An0n3m0us) [reply init.lua]: Oops yes; done!

File: init.lua (hunk 23 above)
Comment (An0n3m0us) [reply init.lua]: I've removed it.

File: init.lua (hunk 18 above)
Comment (octacian) [reply init.lua]: If I understand your question correctly, no, it doesn't need to be implemented as it is. But it does need to be flipped as I mentioned in my first comment. I'll address the other part of this in the comment on line 169.

File: init.lua (hunk 19 above)
Comment (octacian) [reply init.lua:171]: All that is necessary is to define a local player variable, as I have suggested, placing the definition on the line directly above the `if player then` statement, i.e. line 167.

File: init.lua [hunk 24]
```diff
@@ -283,7 +283,7 @@ minetest.register_chatcommand("rank", {
 		if #param == 1 and param[1] == "list" then
 			return true, "Available Ranks: "..ranks.list_plaintext()
 		elseif #param == 2 then
-			if minetest.get_player_by_name(param[1]) then
+			if ranks.get_rank(param[1]) then
```
Comment (octacian) [inline init.lua]: I've just realized this check is probably not needed at all. I'm pretty sure it will always return false for players that don't have a rank yet, so it would be impossible to set the rank of a player with no rank. So yeah, it looks to me as though the entire check could probably be removed.

File: init.lua (hunk 21 above)
Comment (octacian) [reply init.lua]: Oh yeah, this one still does need to be flipped. So
```lua
if type(name) ~= "string" then
    name = name:get_player_name()
end
```
and then define a local player variable assigning `minetest.get_player_by_name(name)` just before the `if player then ... end` statement currently on line 206.

File: init.lua (hunk 24 above)
Comment (An0n3m0us) [reply init.lua]: Fixed

File: init.lua (hunk 21 above)
Comment (An0n3m0us) [reply init.lua]: Okay, done!

File: init.lua (hunk 19 above)
Comment (An0n3m0us) [reply init.lua:171]: Done!

File: init.lua (hunk 18 above)
Comment (An0n3m0us) [reply init.lua]: Done!

**Comment by An0n3m0us** [comment]
I have added backwards compatibility with the database entries.
They will be used if an item does not exist in the new storage method.
And the old item (in the players database) will be renamed, not removed. (We can change this later if needed)

File: init.lua [hunk 25]
```diff
@@ -147,18 +144,19 @@ function ranks.update_privs(player, trigger)
...
 	if minetest.settings:get("ranks.prefix_nametag") == "false" then
 		return
 	end
 
-	if type(player) == "string" then
-		player = minetest.get_player_by_name(player)
+	if type(name) ~= "string" then
+		player = name:get_player_name()
+	else
+		player = minetest.get_player_by_name(name)
```
Comment (octacian) [inline init.lua]: This should be replaced with
```lua
if type(name) ~= "string" then
    name = name:get_player_name()
end
```
as `player` should be defined on line 169 immediately before the if statement where it is necessary.

File: init.lua (hunk 19 above)
Comment (octacian) [reply init.lua:171]: Not quite, `player` still needs to be defined on line 169 as a local variable in this scope, defining it anywhere else would either make it global or just add unnecessary load to fetch the player object so early that it might not be necessary.
Definition should look like this:
```lua
local player = minetest.get_player_by_name(name)
```
(we know that `name` will always be a string because of the if check at the opening of the function)

File: init.lua [hunk 26]
```diff
@@ -169,61 +167,65 @@ function ranks.update_nametag(player)
...
 
 -- [function] Remove rank from player
-function ranks.remove_rank(player)
-	if type(player) == "string" then
-		player = minetest.get_player_by_name(player)
+function ranks.remove_rank(name)
+	if type(name) ~= "string" then
+		player = name:get_player_name()
+	else
+		player = minetest.get_player_by_name(name)
```
Comment (octacian) [inline init.lua]: Like in `update_nametag`, this should be replaced with
```lua
if type(name) ~= "string" then
    name = name:get_player_name()
end
```

File: init.lua [hunk 27]
```diff
@@ -169,61 +167,65 @@ function ranks.update_nametag(player)
...
-		})
-		-- Update privileges
-		local basic_privs =
-			minetest.string_to_privs(minetest.settings:get("basic_privs") or "interact,shout")
-		minetest.set_player_privs(name, basic_privs)
+	local rank = ranks.get_rank(name)
+	if rank ~= nil then
+		storage:set_string(name, nil)
+
+		if player then
```
Comment (octacian) [inline init.lua:212]: ...and `player` should then be defined on the line before this if statement begins (i.e. line 209)
```lua
local player = minetest.get_player_by_name(name)
```

**Review by octacian — COMMENTED** [review]
I think this should be the last round of review! Thanks for sticking with this!
Once the issues I've mentioned are fixed and you can confirm that you've tested and it works, I'll merge the PR.

**Comment by An0n3m0us** [comment]
> I think this should be the last round of review! Thanks for sticking with this!
> Once the issues I've mentioned are fixed and you can confirm that you've tested and it works, I'll merge the PR.

Just before I check over these; I've already had this PR merged into Craigs server, so I guess that might count as testing. Seems to be working okay so far (am I glad! xD)

I'll try have this done by Monday as I am a little busy attending to other things at the moment.

**Comment by octacian** [comment]
Sounds good, I'm glad to hear it's been working alright! With that in mind, I'll merge as soon as you have those fixes committed :)
And no worries if it takes some time, I've been more than a little slow to address this PR over the last few months :smile:

File: init.lua (hunk 25 above)
Comment (An0n3m0us) [reply init.lua]: image

Like so?

Comment (An0n3m0us) [reply init.lua]: Ah, I have seen your next review. Changed it to:
`local player = minetest.get_player_by_name(name)`

**Comment by An0n3m0us** [comment]
It should all be done now. I have tested it too.

**Comment by octacian** [comment]
Alright. Hi. Yeah it's been a while haha. If you've tested and it's all good, I'll take your word for it at this point. Thanks for your help! Merging.
